models/exchange/binance/api.py

```diff
@@ -302,95 +461,171 @@ def marketSell(self, market: str='', base_quantity: float=0) -> list:
-            raise TypeError('Binance market required.')
+                message = f'{method} ({resp.status_code}) {self._api_url}{uri} - {resp_message}'
+                if self.die_on_api_error:
+                    raise Exception(message)
+                else:
+                    Logger.error(f'Error: {message}')
+                    return {}
 
-        return self.client.get_symbol_info(symbol=market)
+            resp.raise_for_status()
+            return resp.json()
+        
+        except requests.ConnectionError as err:
+            return self.handle_api_error(err, 'ConnectionError')
 
+        except requests.exceptions.HTTPError as err:
+            return self.handle_api_error(err, 'HTTPError')
 
-    def getMarketInfoFilters(self, market: str) -> pd.DataFrame:
-        return pd.DataFrame(self.client.get_symbol_info(symbol=market)['filters'])
+        except requests.Timeout as err:
+            return self.handle_api_error(err, 'Timeout')
 
+        except json.decoder.JSONDecodeError as err:
+            return self.handle_api_error(err, 'JSONDecodeError')
 
-    def getTicker(self, market:str) -> tuple:
-        # validates the market is syntactically correct
-        if not self._isMarketValid(market):
-            raise TypeError('Binance market required.')
+    def handle_api_error(self, err: str, reason: str) -> dict:
+        """Handler for API errors"""
 
-        resp = self.client.get_symbol_ticker(symbol=market)
+        if self.debug:
+            if self.die_on_api_error:
+                raise SystemExit(err)
+            else:
+                Logger.debug(err)
+                return {}
+        else:
+            if self.die_on_api_error:
+                raise SystemExit(f"{reason}: {self._api_url}")
+            else:
+                Logger.info(f"{reason}: {self._api_url}")
+                return {}
 
-        if 'price' in resp:
-            return (self.getTime().strftime('%Y-%m-%d %H:%M:%S'), float('{:.8f}'.format(float(resp['price']))))
+class PublicAPI(AuthAPIBase):
+    def __init__(self, api_url='https://api.binance.com') -> None:
+        """Binance API object model
+    
+        Parameters
+        ----------
+        api_url
+            Binance API URL
+        """
 
-        now = datetime.today().strftime('%Y-%m-%d %H:%M:%S')
-        return (now, 0.0)
+        # options
+        self.debug = False
+        self.die_on_api_error = False
+
+        valid_urls = [
+            'https://api.binance.com',
+            'https://api.binance.us',
+            'https://testnet.binance.vision'
+        ]
+
+        # validate Binance API
+        if api_url not in valid_urls:
+            raise ValueError('Binance API URL is invalid')
 
+        self._api_url = api_url
 
     def getTime(self) -> datetime:
         """Retrieves the exchange time"""
     
+        def convert_time(epoch: int=0):
+            epoch_str = str(epoch)[0:10]
+            return datetime.fromtimestamp(int(epoch_str))
```

Comment:
maybe we could move this method into AuthAPIBase, which all other classes where it's defined inherit from. So it would be defined only once, and available in the subclasses?

Reply:
Yes, sounds like a good idea.

Reply:
This is done, thanks.